daffodil-runtime1/src/main/java/org/apache/daffodil/udf/UDFunctionService.java

```diff
@@ -0,0 +1,137 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.udf;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.ServiceConfigurationError;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+
+import org.apache.daffodil.util.Misc;
+
+public class UDFunctionService {
+	private ServiceLoader<UDFunctionProvider> loader;
+	private ArrayList<Class<?>> functionClasses = new ArrayList<>();
+	private HashMap<String, UDFunctionProvider> functionProviderLookup = new HashMap<>();
+	private ArrayList<String> errors = new ArrayList<>();
+	private ArrayList<String> warnings = new ArrayList<>();
+
+	public UDFunctionService(ClassLoader cl) {
+		try {
+			loader = cl == null ? ServiceLoader.load(UDFunctionProvider.class) : ServiceLoader.load(UDFunctionProvider.class, cl);
+
+			loader.forEach(fcp -> {
+				Class<?>[] fcpfc = fcp.getFunctionClasses();
+
+				if (fcpfc == null || fcpfc.length == 0) {
+					warnings.add(String.format("Provider ignored: [%s].\nNo Function Classes found.",
+							fcp.getClass().getName()));
+					return;
+				}
+
+				List<Class<?>> fc = Arrays.asList(fcpfc);
+
+				List<Class<?>> missingAnnotations = fc.stream()
+						.filter(f -> !f.isAnnotationPresent(FunctionClassInfo.class)).collect(Collectors.toList());
+
+				List<Class<?>> nonSerializables = fc.stream().filter(f -> Serializable.class.isAssignableFrom(f))
+						.collect(Collectors.toList());
+
+				if (!missingAnnotations.isEmpty()) {
+					warnings.add(
+							String.format("Provider ignored: [%s].\nAnnotations missing for FunctionClass(es):\n%s",
+									fcp.getClass().getName(), missingAnnotations.stream().map(c -> c.getName())
+											.collect(Collectors.joining("\n", "\t[", "]"))));
+					return;
+				}
+
+				if (!nonSerializables.isEmpty()) {
+					warnings.add(String.format(
+							"Provider ignored: [%s].\nFunctionClass(es) must implement java.io.Serializable:\n%s",
+							fcp.getClass().getName(), nonSerializables.stream().map(c -> c.getName())
+									.collect(Collectors.joining("\n", "\t[", "]"))));
+					return;
+				}
+
+				fc.stream().forEach(f -> {
+					FunctionClassInfo fInfo = f.getAnnotation(FunctionClassInfo.class);
+					String ns = fInfo.namespace();
+					String fname = fInfo.name();
+					if (ns.isEmpty() || fname.isEmpty() || ns.matches("\\s+") || fname.matches("\\s+")) {
```

Comment:
Instead of the matches, you can probably do ns.trim().isEmpty() || fname.trim().isEmpty()